pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,472 @@
+		f.intentAccessors.Lock(image)
+		defer f.intentAccessors.Unlock(image)
+
+		var err error
+		var exists bool
+		pulledRecord, exists, err = f.recordsAccessor.GetImagePulledRecord(imageRef)
+		switch {
+		case err == nil && pulledRecord == nil:
+			if exists, err := f.recordsAccessor.ImagePullIntentExists(image); err != nil {
+				return fmt.Errorf("failed to check existence of an image pull intent: %w", err)
+			} else if exists {
+				imagePulledByKubelet = true
+			}
+		case err == nil && pulledRecord != nil:
+			imagePulledByKubelet = true
+		case err != nil && exists: // record exists but has invalid format
+		default:
+			return fmt.Errorf("failed to retrieve image pulled record: %w", err)
+		}
+
+		return nil
+	}()
+
+	if err != nil {
+		klog.ErrorS(err, "Unable to access cache records about image pulls")
+		return true
+	}
+
+	if !f.requireCredentialVerification(image, len(imageRef) > 0, imagePulledByKubelet) {
+		return false
+	}
+
+	if pulledRecord == nil {
+		// we have no proper records of the image being pulled in the past, we can short-circuit here
+		return true
+	}
+
+	sanitizedImage, err := trimImageTagDigest(image)
+	if err != nil {
+		klog.ErrorS(err, "failed to parse image name, forcing image credentials reverification", "image", sanitizedImage)
+		return true
+	}
+
+	cachedCreds, ok := pulledRecord.CredentialMapping[sanitizedImage]
+	if !ok {
+		return true
+	}
+
+	if cachedCreds.NodePodsAccessible {
+		// anyone on this node can access the image
+		return false
+	}
+
+	if len(cachedCreds.KubernetesSecrets) == 0 {
+		return true
+	}
+
+	for _, podSecret := range podSecrets {
+		for _, cachedSecret := range cachedCreds.KubernetesSecrets {
+
+			// we need to check hash len in case hashing failed while storing the record in the keyring
+			hashesMatch := len(cachedSecret.CredentialHash) > 0 && podSecret.CredentialHash == cachedSecret.CredentialHash
+			secretCoordinatesMatch := podSecret.UID == cachedSecret.UID &&
+				podSecret.Namespace == cachedSecret.Namespace &&
+				podSecret.Name == cachedSecret.Name
+
+			if hashesMatch {
+				if !secretCoordinatesMatch && len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit {
+					// While we're only matching at this point, we want to ensure this secret is considered valid in the future
+					// and so we make an additional write to the cache.
+					// writePulledRecord() is a noop in case the secret with the update hash already appears in the cache.
+					if err := f.writePulledRecord(image, imageRef, &kubeletconfiginternal.ImagePullCredentials{KubernetesSecrets: []kubeletconfiginternal.ImagePullSecret{podSecret}}); err != nil {
+						klog.ErrorS(err, "failed to write an image pulled record", "image", image, "imageRef", imageRef)
+					}
+				}
+				return false
+			}
+
+			if secretCoordinatesMatch {
+				if len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit { // we already know the hashes did not match, no need to check again
```

Comment:
let's just make this symmetrical so if the code moves it is still correct / efficient, and drop the need for a comment

```suggestion
				if !hashesMatch && len(cachedCreds.KubernetesSecrets) < writeRecordWhileMatchingLimit {
```